packages/flutter_tools/lib/src/features.dart

```diff
@@ -0,0 +1,245 @@
+    enabledByDefault: false,
+  ),
+);
+
+/// The [Feature] for Windows desktop.
+const Feature flutterWindowsDesktopFeature = Feature(
+  name: 'Flutter Desktop for Windows',
+  configSetting: 'enable-windows-desktop',
+  environmentOverride: 'ENABLE_FLUTTER_DESKTOP',
+  master: FeatureSetting(
+    available: true,
+    enabledByDefault: false,
+  ),
+);
+
+/// A [FeatureFlags] that looks up values based on [Feature] definitions.
+class FlutterFeatureFlags implements FeatureFlags {
+  const FlutterFeatureFlags();
+
+  @override
+  bool get isLinuxEnabled => _isEnabled(flutterLinuxDesktopFeature);
+
+  @override
+  bool get isMacOSEnabled => _isEnabled(flutterMacOSDesktopFeature);
+
+  @override
+  bool get isWebEnabled => _isEnabled(flutterWebFeature);
+
+  @override
+  bool get isWindowsEnabled => _isEnabled(flutterWindowsDesktopFeature);
+
+  // Calculate whether a particular feature is enabled for the current channel.
+  static bool _isEnabled(Feature feature) {
+    final String currentChannel = FlutterVersion.instance.channel;
+    final FeatureSetting featureSetting = feature.getSettingForChannel(currentChannel);
+    if (!featureSetting.available) {
+      return false;
+    }
+    bool isEnabled = featureSetting.enabledByDefault;
+    if (feature.configSetting != null) {
+      final bool configOverride = Config.instance.getValue(feature.configSetting);
+      if (configOverride != null) {
+        isEnabled = configOverride;
+      }
+    }
+    if (feature.environmentOverride != null) {
+      if (platform.environment[feature.environmentOverride]?.toLowerCase() == 'true') {
+        isEnabled = true;
+      }
+    }
+    return isEnabled;
+  }
+}
+
+/// A [Feature] is a process for conditionally enabling tool features.
+///
+/// All settings are optional, and if not provided will generally default to
+/// a "safe" value, such as being off.
+///
+/// The top level feature settings can be provided to apply to all channels.
+/// Otherwise, more specific settings take precidence over higher level
+/// settings.
+class Feature {
+  /// Creates a [Feature].
+  const Feature({
+    @required this.name,
+    this.environmentOverride,
+    this.configSetting,
+    this.master = const FeatureSetting(),
+    this.dev = const FeatureSetting(),
+    this.beta = const FeatureSetting(),
+    this.stable = const FeatureSetting(),
+  });
+
+  /// The user visible name for this feature.
+  final String name;
+
+  /// The settings for the master branch and other unknown channels.
+  final FeatureSetting master;
+
+  /// The settings for the dev branch.
+  final FeatureSetting dev;
+
+  /// The settings for the beta branch.
+  final FeatureSetting beta;
+
+  /// The settings for the stable branch.
+  final FeatureSetting stable;
+
+  /// The name of an environment variable that can override the setting.
+  ///
+  /// The environment variable only needs to be "set", that is contain a
+  /// non empty string. This is only intended for usage by CI and not
```

Comment:
that's not what you implemented on line 135. Add a test

Reply:
Done